### PR TITLE
Added support for type based parameter binding

### PR DESCRIPTION
**Pull request by davidfowl**
- Added a convention that allows custom async binding logic to run for parameters that have a static BindAsync method that takes an HttpContext and return a ValueTask of object. This allows customers to write custom binders based solely on type (it's an extension of the existing TryParse pattern).
- There's allocation overhead per request once there's a parameter binder for a delegate. This is because we need to box all of the arguments since we're not using generated code to compute data from the list of binders.
- This also supports optionality like everything else since we're always boxing using object.
- Changed TryParse tests to BindAsync tests and added more tests.

Usage:

```C#
var builder = WebApplication.CreateBuilder(args);
var app = builder.Build();

app.MapGet("/webhook", (CloudEvent cloudEvent) =>
{
    redis.Publish(cloudEvent);
});

app.Run();

public class CloudEvent
{
    public static async ValueTask<object?> BindAsync(HttpContext context)
    {
        return await CloudEventParser.ReadEventAsync(context)
    }
}
```

### CODE REVIEW

File: src/Http/Http.Extensions/src/RequestDelegateFactory.cs [hunk 1]
```diff
@@ -275,7 +274,7 @@ private static Expression CreateArgument(ParameterInfo parameter, FactoryContext
                 // when RDF.Create is manually invoked.
                 if (factoryContext.RouteParameters is { } routeParams)
                 {
-                   
+
```
Comment (davidfowl) [inline src/Http/Http.Extensions/src/RequestDelegateFactory.cs:277]:
```suggestion
```

**Review by halter73 — APPROVED** [review]

**Comment by JesperTreetop** [comment]
Now that `static` members are allowed in interfaces, this kind of thing could be specified as an interface: ([sharplab](https://sharplab.io/#v2:EYLgtghglgdgNAExAagD4AEBMBGAsAKHQAYACdbAVgG4CCB5YAKwFMBjAFwH4SAhWBAIIBnAJ4xWAHgAqAPgAUACXbsADgGEA9jHbMAHuxKstO/QEoSAdwAWzAE7MSUkiBIBJKSJXM+MQaPEkAN4EJKFkAOyOAHQ+fmKsckbaeuymNPgAvrT42SFh+QUEWI6e3vzC8RDAADYOLu6lsRUBwfgF6ADMZNgAbGQAHCQAahDVAK7M6D0SGkxsXDK85f4JSqqayfqGxinmALyLMMwWJLMsHHJpBFk5hF2wOrYAZhCsDg1eTStBeaHkfVUhOxbK8DCNxpNpmd5pxFl94oplOodlskiZUukbkUusU1sjNgZAhkgA===))

```csharp
public interface ITypeBindAsync {
    static abstract ValueTask<object?> BindAsync(HttpContext context);
}
```

It wouldn't make it easier to call, though, since you can't exactly go `Type type; if (type is ITypeBindAsync T) { return T.BindAsync(context); }`. (Making it have a "[curiously recurring](https://devblogs.microsoft.com/dotnet/preview-features-in-net-6-generic-math/)" type parameter which almost all `static abstract` examples involve, ie giving it a `TSelf`, also wouldn't make it easier to call.)

But it would be easier and probably quicker to find out if a type is compliant or not - look for the type implementing the interface instead of looking for the method implementation. Then again, you still have to have the method to actually *call* it the way this is implemented now, through expression tree compilation. And of course, if you wanted the pattern to be usable without implementing an interface, that change couldn't be made.

Basically, the patch is probably the best way to do it, but it's fun to finally be able to think about statics in interfaces!
